rbc/omnisci_backend/omnisci_buffer.py

```diff
@@ -0,0 +1,296 @@
+"""Implement Buffer type as a base class to Omnisci Array and Column types.
+
+Omnisci Buffer represents the following structure:
+
+  template<typename T>
+  struct Buffer {
+    T* ptr;
+    size_t sz;
+    ...
+  }
+
+that is, a structure that has at least two members where the first is
+a pointer to some data type and the second is the size of the buffer.
+
+This module implements the support for the following Python operators:
+
+  len
+  __getitem__
+  __setitem__
+
+to provide a minimal functionality for accessing and manipulating the
+Omnisci buffer objects from UDF/UDTFs.
+"""
+
+
+import operator
+from llvmlite import ir
+from rbc import typesystem
+from rbc.utils import get_version
+if get_version('numba') >= (0, 49):
+    from numba.core import datamodel, cgutils, extending, types
+else:
+    from numba import datamodel, cgutils, extending, types
+
+
+int8_t = ir.IntType(8)
+int32_t = ir.IntType(32)
+int64_t = ir.IntType(64)
+void_t = ir.VoidType()
+
+
+class OmnisciBufferType(typesystem.Type):
+
+    @property
+    def _buffer_typename(self):
+        return self._params['typename']
+
+    def pointer(self):
+        ptr_type = self._params.get('pointer')
+        if ptr_type is not None:
+            return ptr_type
+        return typesystem.Type.pointer(self)
+
+
+class BufferType(types.Type):
+
+    pass
+
+
+class BufferPointer(types.Type):
+    """Type class for pointers to Omnisci buffer structures.
+
+    We are not deriving from CPointer because BufferPointer getitem is
+    used to access the data stored in Buffer ptr member.
+    """
+    mutable = True
+
+    def __init__(self, dtype, eltype):
+        self.dtype = dtype    # struct dtype
+        self.eltype = eltype  # buffer element dtype
+        name = "(%s)*" % dtype
+        types.Type.__init__(self, name)
+
+    @property
+    def key(self):
+        return self.dtype
+
+
+class Buffer(object):
+    """Represents Omnisci Buffer that can be constructed within UDF/UDTFs.
+    """
+
+
+class BufferPointerModel(datamodel.models.PointerModel):
+    pass
+
+
+@extending.intrinsic
+def omnisci_buffer_ptr_len_(typingctx, data):
+    sig = types.int64(data)
+
+    def codegen(context, builder, signature, args):
+        data, = args
+        rawptr = cgutils.alloca_once_value(builder, value=data)
+        struct = builder.load(builder.gep(rawptr,
+                                          [int32_t(0)]))
+        return builder.load(builder.gep(
+            struct, [int32_t(0), int32_t(1)]))
+    return sig, codegen
+
+
+@extending.intrinsic
+def omnisci_buffer_len_(typingctx, data):
+    sig = types.int64(data)
+
+    def codegen(context, builder, signature, args):
+        data, = args
+        assert data.opname == 'load'
```

Comment:
Is this assertion required?

Reply:
I am uncertain. The subsequent `buf = data.operands[0]` relies on that `data.opname == 'load'`. If the opname is not `load`, using `operands[0]` will lead to unpredictable behaviour and it seems to me that tracking down this possible(?) scenario will be very difficult. So I think the assert statement can be removed if one can be certain that opname is `load`.